x86_symbolic/src/Data/Macaw/X86/Symbolic.hs

```diff
@@ -472,6 +480,102 @@ instance GenArchInfo LLVMMemory M.X86_64 where
     , updateReg = x86UpdateReg
     }
 
+-- | Helper, not exported
+writeBytesBackwards ::
+  MM.HasPtrWidth w =>
+  C.IsSymBackend sym bak =>
+  (?memOpts :: MM.MemOptions) =>
+  MM.HasLLVMAnn sym =>
+  bak ->
+  MM.MemImpl sym ->
+  MM.LLVMPtr sym w ->
+  Seq.Seq Word8 ->
+  IO (MM.LLVMPtr sym w, MM.MemImpl sym)
+writeBytesBackwards bak mem ptr bytes = do
+  let sym = C.backendGetSym bak
+  let i8 = MM.bitvectorType (Bytes.toBytes (1 :: Int))
+  z <- WI.natLit sym 0
+  one <- WI.bvOne sym ?ptrWidth
+  let writeByte (p, m) byte = do
+        p' <- MM.ptrSub sym ?ptrWidth p one
+        bv <- WI.bvLit sym (C.knownNat @8) (BVS.mkBV C.knownNat (fromIntegral byte))
+        m' <- MM.storeRaw bak m p' i8 CLD.noAlignment (MM.LLVMValInt z bv)
+        pure (p', m')
+  Monad.foldM writeByte (ptr, mem) (Seq.reverse bytes)
+
+-- | Create an allocation for the stack for x86_64 with the SysV ABI.
+--
+-- On x86_64 with the SysV ABI, the stack grows \"downwards\" from high
+-- addresses to low. The end of the stack is initialized with the ELF auxiliary
+-- vector (not modelled here), and functions expect the following data to be
+-- available above their stack frame (i.e., just above the address in @rsp@):
+--
+-- * Their stack-spilled arguments
+-- * The return address
+--
+-- To model this behavior, this function:
+--
+-- 1. Uses 'MSS.createArrayStack' to allocate the stack, which returns a stack
+--    allocation and a pointer to its end, minus the space for stack-spilled
+--    arguments (see its Haddocks for details)
+-- 2. Decrements the pointer to the top of the stack ('MSS.arrayStackTopPtr') by
+--    8 bytes (i.e., the size of a pointer)
+-- 3. Writes a symbolic return address to the top of the stack
+--
+-- (Unless a concrete list of bytes is passed, in which case the decrement (2)
+-- is by the length of that list, and the write (3) is of those bytes.)
+--
+-- This setup should enable calling most functions without incident, after the
+-- returned pointer is put in @rsp@. Those that expect stack-spilled arguments
+-- may end up reading from the (uninitialized) stack slots.
+--
+-- Notably, the Microsoft x86_64 ABI differs in the stack setup. See
+-- [this link](https://learn.microsoft.com/en-us/cpp/build/stack-usage)
+-- for more detailson that ABI.
+x86SysvStack ::
+  C.IsSymBackend sym bak =>
+  (?memOpts :: MM.MemOptions) =>
+  MM.HasLLVMAnn sym =>
+  bak ->
+  MM.MemImpl sym ->
+  -- | Initialize the end of the stack to a concrete sequence of bytes if the
+  -- value is @Just@, otherwise initialize the end of the stack to 8 fresh,
+  -- symbolic bytes.
+  Maybe (Seq.Seq Word8) ->
```

Comment:
I believe it is always the case that the return address is a 64-bit word. If so, it might be better to model this as a `BV 64` instead of as a `Seq Word8` to avoid unrepresentable states.

Reply:
Yep, I'm refactoring all this, and plan to do this.